comm/src/main/scala/coop/rchain/comm/rp/HandleMessages.scala

```diff
@@ -73,22 +76,43 @@ object HandleMessages {
       _ <- ConnectionsCell[F].addConnAndReport(peer)
     } yield handledWithoutMessage
 
-  def handleProtocolHandshake[F[_]: Monad: TransportLayer: Log: ConnectionsCell: RPConfAsk: Metrics](
+  def handleProtocolHandshake[F[_]: Monad: Sync: TransportLayer: Log: ConnectionsCell: RPConfAsk: Metrics](
       peer: PeerNode,
       protocolHandshake: ProtocolHandshake
-  ): F[CommunicationResponse] =
-    for {
-      conf     <- RPConfAsk[F].ask
-      response = ProtocolHelper.protocolHandshakeResponse(conf.local, conf.networkId)
-      resErr   <- TransportLayer[F].send(peer, response)
-      _ <- resErr.fold(
-            kp(().pure[F]),
-            kp(
-              Log[F].info(s"Responded to protocol handshake request from $peer") >>
-                ConnectionsCell[F].addConnAndReport(peer)
+  ): F[CommunicationResponse] = {
+    val confF = RPConfAsk[F].ask
+    def canAddNewHost: F[Boolean] =
+      for {
+        conf                <- confF
+        isCurrentPeerLocal  <- isValidPublicInetAddress[F](conf.local.endpoint.host)
+        isIncomingPeerLocal <- isValidPublicInetAddress[F](peer.endpoint.host)
+        bothPeersAreLocal   = isCurrentPeerLocal && isIncomingPeerLocal
+        bothPeersArePublic  = !isCurrentPeerLocal && !isIncomingPeerLocal
+      } yield (bothPeersAreLocal || bothPeersArePublic)
+
+    def acceptConnection: F[CommunicationResponse] =
+      for {
+        conf     <- confF
+        response = ProtocolHelper.protocolHandshakeResponse(conf.local, conf.networkId)
+        resErr   <- TransportLayer[F].send(peer, response)
+        _ <- resErr.fold(
+              kp(().pure[F]),
+              kp(
+                Log[F].info(s"Responded to protocol handshake request from $peer") >>
+                  ConnectionsCell[F].addConnAndReport(peer)
+              )
             )
-          )
-    } yield handledWithoutMessage
+      } yield handledWithoutMessage
+
+    def declineConnection: F[CommunicationResponse] = {
+      val _ = Log[F].info(
+        s"Unable to add connection: incoming address ${peer.toAddress}"
+      )
+      handledWithoutMessage.pure[F]
+    }
+
+    canAddNewHost.ifM(acceptConnection, declineConnection)
```

Comment:
It is possible that value inside RPConfAsk change between calls in `canAddNewHost` and `acceptConnection`.
To ensure that you work with the same value, first extract the conf value and supply it to both `canAddNewHost` and `acceptConnection` functions.